### DISCUSSION TITLE
How to make Generator.IdentifierName not add @ or is there any other way to replace it?

### DISCUSSION THREAD

**Discussion by yika-aixi** [dotnet/roslyn#50904] [Q&A]
Roslyn Ver: 3.8.0-5.final

I am make a code generation library. When dealing with the common convention WithTypeConstraint, IdentifierName will add @ to the structure. What should I do? Is there any alternative?

Below is my code:

```

 internal readonly SyntaxGenerator Generator;

----

public SyntaxNode GetSyntaxNode(CodeFile file)
{
	return file.Generator.IdentifierName(Value);
}

-----

private Dictionary<string,GenericWhereItem[]> _wheres = new Dictionary<string, GenericWhereItem[]>();
/* elements
  ["T",
    ["struct"]
  ]
*/
foreach (var @where in _wheres)
{
    declaration = generator.WithTypeConstraint(declaration, where.Key, where.Value.Select(x=> x.GetSyntaxNode(file)).ToArray());
}
```

Generated code:

![image](https://user-images.githubusercontent.com/22412994/106332691-41ad8380-62c2-11eb-9ba1-f6017006ed4e.png)

**Comment by CyrusNajmabadi** [comment]
> I am make a code generation library. When dealing with the common convention WithTypeConstraint, IdentifierName will add @ to the structure. 

That is correct.   You using an identifier there, so it has to be escaped in order to not be interpretted as the keyword.  If you want the actual `struct` constraint, use `SpecialTypeConstraintKind.ValueType` instead.

**Reply by yika-aixi** [reply]
Is there no other way except him?

**Reply by CyrusNajmabadi** [reply]
You don't have you use SyntaxGenerator. You can just manually construct the node with SyntaxFactory

**Reply by yika-aixi** [reply]
Oh, I have changed it to the way you said, used SpecialTypeConstraintKind

Thank you
